q1043/q1043_recursive.py

```diff
@@ -0,0 +1,28 @@
+class Solution:
+    def maxSumAfterPartitioning(self, arr: List[int], k: int) -> int:
+        memo = [-1 for _ in range(len(arr))]
```

Comment:
memo = defaultdict(int) could also be used to handle dummy assignments. In that case, line 12 has to be 
`if start in memo:` But your approach is also good.

Reply:
I was also looking at this same problem, coincidentally :)

Reply:
Totally agree! It's a nice thing that python has, but I use an array for memo mostly because of the following two reasons:
- An array has better support across different languages, defaultdict only exists in Python 🥲
- In cases like this, I think a simple array can be more memory efficient. 

But I like defaultdict a lot too. It's so convenient.

Reply:
I see. Good to know.